awx/network_ui/consumers.py

```diff
@@ -0,0 +1,277 @@
+                                   from_device_id=device_map[link['from_device_id']],
+                                   to_device_id=device_map[link['to_device_id']],
+                                   from_interface_id=Interface.objects.get(device_id=device_map[link['from_device_id']],
+                                                                           id=link['from_interface_id']).pk,
+                                   to_interface_id=Interface.objects.get(device_id=device_map[link['to_device_id']],
+                                                                         id=link['to_interface_id']).pk)
+        (Topology.objects
+                 .filter(topology_id=topology_id, link_id_seq__lt=link['id'])
+                 .update(link_id_seq=link['id']))
+
+    def onLinkDestroy(self, link, topology_id, client_id):
+        device_map = dict(Device.objects
+                                .filter(topology_id=topology_id, id__in=[link['from_device_id'], link['to_device_id']])
+                                .values_list('id', 'pk'))
+        Link.objects.filter(id=link['id'],
+                            from_device_id=device_map[link['from_device_id']],
+                            to_device_id=device_map[link['to_device_id']],
+                            from_interface_id=Interface.objects.get(device_id=device_map[link['from_device_id']],
+                                                                    id=link['from_interface_id']).pk,
+                            to_interface_id=Interface.objects.get(device_id=device_map[link['to_device_id']],
+                                                                  id=link['to_interface_id']).pk).delete()
+
+    def onDeviceSelected(self, message_value, topology_id, client_id):
+        'Ignore DeviceSelected messages'
+        pass
+
+    def onDeviceUnSelected(self, message_value, topology_id, client_id):
+        'Ignore DeviceSelected messages'
+        pass
+
+    def onLinkSelected(self, message_value, topology_id, client_id):
+        'Ignore LinkSelected messages'
+        pass
+
+    def onLinkUnSelected(self, message_value, topology_id, client_id):
+        'Ignore LinkSelected messages'
+        pass
+
+    def onMultipleMessage(self, message_value, topology_id, client_id):
+        for message in message_value['messages']:
+            handler = self.get_handler(message['msg_type'])
+            if handler is not None:
+                handler(message, topology_id, client_id)
+            else:
+                logger.warning("Unsupported message %s", message['msg_type'])
+
+
+persistence = _Persistence()
+
+
+# UI Channel Events
+
+@channel_session
+def ws_connect(message):
+    # Accept connection
+    data = urlparse.parse_qs(message.content['query_string'])
+    inventory_id = parse_inventory_id(data)
+    topology_ids = list(TopologyInventory.objects.filter(inventory_id=inventory_id).values_list('topology_id', flat=True))
+    topology_id = 0
```

Comment:
This code assumes that a valid `topology_id` will never be 0, it would be slightly better to set it to `None` and check `if topology_id is not None` on 203

Reply:
Yes,  the database does not restrict ids to positive numbers.  I'll change this.